### ISSUE TITLE
Clippy with `--no-deps` emits unexpected warnings from deps or should have emitted warnings as expected

### ISSUE THREAD

**Issue by zjp-CN** [rust-lang/rust-clippy#13983]
### Summary

Zulipchat: https://rust-lang.zulipchat.com/#narrow/channel/257328-clippy/topic/unexpected.20results.20of.20clippy.20with.20.60--no-deps.60/near/459710102

### Reproducer

There is a minimum reproducible repo: https://github.com/os-checker/bug-MRE-clippy-no-deps

```sh
echo Bug Case 1

cargo clean
echo expected warning
cargo clippy --no-deps --manifest-path ./a-dep/Cargo.toml
echo unexpected warning from a-dep
cargo clippy --no-deps --manifest-path ./b-dep/Cargo.toml

echo Bug Case 2

cargo clean
echo expected no warning from a-dep
cargo clippy --no-deps --manifest-path ./b-dep/Cargo.toml
echo unexpected no warning
cargo clippy --no-deps --manifest-path ./a-dep/Cargo.toml
```

```rust
// a-dep/src/lib.rs
pub struct A {}

impl A {
    pub fn new() -> A {
        A {}
    }
}
```

```rust
/// b-dep/src/lib.rs
```

```rust
$ sh ./reproducible.sh
Bug Case 1
     Removed 23 files, 56.1KiB total
expected warning
    Checking a-dep v0.1.0 (/rust/tmp/clippy-no-deps/a-dep)
warning: you should consider adding a `Default` implementation for `A`
...

warning: `a-dep` (lib) generated 1 warning (run `cargo clippy --fix --lib -p a-dep` to apply 1 suggestion)
    Finished `dev` profile [unoptimized + debuginfo] target(s) in 0.13s
unexpected warning from a-dep 👈👈👈 only check b-dep, but a-dep's result is included here
warning: you should consider adding a `Default` implementation for `A`
...
warning: `a-dep` (lib) generated 1 warning (run `cargo clippy --fix --lib -p a-dep` to apply 1 suggestion) 
    Checking b-dep v0.1.0 (/rust/tmp/clippy-no-deps/b-dep) 👈👈👈 b-dep is empty without any code!
    Finished `dev` profile [unoptimized + debuginfo] target(s) in 0.07s
Bug Case 2
     Removed 24 files, 102.4KiB total
expected no warning from a-dep
    Checking a-dep v0.1.0 (/rust/tmp/clippy-no-deps/a-dep)
    Checking b-dep v0.1.0 (/rust/tmp/clippy-no-deps/b-dep)
    Finished `dev` profile [unoptimized + debuginfo] target(s) in 0.15s
unexpected no warning 👈👈👈 only check a-dep, but no result????
    Finished `dev` profile [unoptimized + debuginfo] target(s) in 0.01s
```

### Version

```text
rustc 1.85.0-nightly (7f75bfa1a 2024-12-30)
binary: rustc
commit-hash: 7f75bfa1ad4e9a9d33a179a90603001515e91991
commit-date: 2024-12-30
host: x86_64-unknown-linux-gnu
release: 1.85.0-nightly
LLVM version: 19.1.6
```

**Comment by blyxyas** [comment]
As a tip for solvers:

AFAICT this comes from Cargo storing diagnostics in `target` and rustc not differentiating if we're on a `--no-deps` when reading those cached diagnostics. If you insert `cargo clean` in between every `cargo` execution it will behave as expected.

So this is more of a reporting more than needed issue, rather than a linting more than needed issue. Still a tricky issue fit for those willing to delve deep in the Rust project.

@rustbot label +E-medium

**Comment by 2asoft** [comment]
I looked into this some, dumping notes here in case they may be useful.

Bisecting shows that the second unexpected scenario (`echo unexpected no warning`) is no longer failing as of https://github.com/rust-lang/rust-clippy/pull/14650.  The exact reason is not clear to me -- the PR is a large change.

---

# Findings

## 1. Location where Clippy applies `--no-deps` filtering

In `src/tools/clippy/src/driver.rs`, lines 307-335:

- `no_deps` is set to `true` if `--no-deps` is present in `CLIPPY_ARGS`
- `relevant_package = !no_deps || env::var("CARGO_PRIMARY_PACKAGE").is_ok()`
- `clippy_enabled = !cap_lints_allow && relevant_package && !info_query`
- If `clippy_enabled` is false, Clippy uses `RustcCallbacks` instead of `ClippyCallbacks`, preventing Clippy lints from being registered

## 2. How this condition is fulfilled by Cargo on non-cached run

In `src/tools/cargo/src/cargo/core/compiler/mod.rs`, lines 770-774:

- For primary packages: `CARGO_PRIMARY_PACKAGE` is set to "1"
- For dependencies: `CARGO_PRIMARY_PACKAGE` is not set
- So, when `--no-deps` is used:
  - Primary packages: `relevant_package = true` (clippy enabled)
  - Dependencies: `relevant_package = false` (clippy disabled)

## 3. How this condition is fulfilled by Cargo on cached run

In `src/tools/cargo/src/cargo/core/compiler/mod.rs`, lines 214-228:

- When a unit is cached, `replay_output_cache` is called instead of compiling
- `replay_output_cache` reads cached diagnostics from a file and replays them
- The cached diagnostics include Clippy warnings emitted during the original compilation
- During replay, the Clippy driver logic is not executed, so `clippy_enabled` is not re-evaluated
- Cached Clippy warnings are replayed regardless of current `--no-deps` setting


During a cached replay, the Clippy driver logic is not executed because Cargo bypasses the compiler entirely.

## Normal (Non-Cached) Compilation Flow
1. **Cargo decides to compile a unit** → calls `rustc`/`clippy-driver` with environment variables set
2. **Clippy driver starts** → `main()` in `src/tools/clippy/src/driver.rs` executes
3. **Environment parsing** → reads `CLIPPY_ARGS`, checks for `--no-deps`
4. **Primary package check** → examines `CARGO_PRIMARY_PACKAGE` environment variable
5. **Decision logic** → computes `relevant_package = !no_deps || CARGO_PRIMARY_PACKAGE.is_ok()`
6. **Lint registration** → if `clippy_enabled`, registers Clippy lints via `ClippyCallbacks`
7. **Compilation & caching** → Rustc runs, emits diagnostics, Cargo caches them in `message_cache_path`

## Cached Replay Flow  
1. **Cargo detects unit is fresh** → skips compilation entirely
2. **Direct replay** → calls `replay_output_cache()` function
3. **File reading** → reads cached diagnostics from disk (bypassing compiler)
4. **Direct emission** → prints cached diagnostics to stderr without any filtering
5. **No driver execution** → `clippy-driver` binary never runs, so no logic in `driver.rs` executes

## The Problem
- **Cached diagnostics contain Clippy warnings** from when the unit was originally compiled with Clippy enabled
- **During replay, no code runs** to re-evaluate whether Clippy should be enabled for the current context
- **Environment variables** like `CARGO_PRIMARY_PACKAGE` may be different during replay
- **No opportunity** to apply the `--no-deps` filtering logic

# Next
cargo doesn't really have the notion of "--no-deps", this is a clippy-only feature, and so it makes sense to let clippy do the filtering.  Cached warnings do have a prefix of "clippy::", which could be used to detect which "driver" to let perform the filtering.

**Comment by km274** [comment]
@2asoft, I'll just throw it out there that you're welcome to the issue if you were planning to claim it. It sounds like you put some good time in, and I haven't started aside from reading the issue summary itself.
